### PR TITLE
[enhancement] simplify array_api enabling tags via wrapper

### PR DESCRIPTION
**Pull request by icfaust**
## Description

Add ```enable_array_api``` wrapper for classes.  This will set the necessary method (```__sklearn_tags__``` for sklearn>=1.6 or ```_more_tags``` for sklearn > 1.3) to signify to the sklearnex infrastructure that ```array_api_dispatch``` would work with onedal for that estimator when enabled.  Otherwise this code would have to be manually added to each estimator, this will simplify development.

No performance benchmarks necessary as it is an import time cost.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: sklearnex/utils/_array_api.py [hunk 1]
```diff
@@ -80,3 +85,38 @@ def get_namespace(*arrays):
...
+        Estimator class.
+    """
+    if sklearn_check_version("1.6"):
+
+        def __sklearn_tags__(self) -> Tags:
+            sktags = super(self, original_class).__sklearn_tags__()
+            sktags.onedal_array_api = True
+            return sktags
+
+        original_class.__sklearn_tags__ = __sklearn_tags__
```
Comment (david-cortes-intel) [inline sklearnex/utils/_array_api.py:113]: Wouldn't modification of a class like this prevent serialization with pickle afterwards? How about making it a decorator?

Comment (icfaust) [reply sklearnex/utils/_array_api.py:113]: Could you provide more info on the serialization issue? won't the environment require a definition for that class method in order to load (and thereby use the class wrapper (something like: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/daal4py/sklearn/_n_jobs_support.py#L208)? For sklearn < 1.6 the ```_get_tags``` attribute acts a bit odd: https://github.com/scikit-learn/scikit-learn/blob/1.3.X/sklearn/base.py#L363 where it will traverse the classes in the mro in a very specific fashion.  For the __sklearn_tags__ attribute, I think your suggestion should be good to prevent weirdness where someone modifies tags specifically in the sklearnex class definition. Let me know @david-cortes-intel

Comment (david-cortes-intel) [reply sklearnex/utils/_array_api.py:113]: Looks like it's not a problem anymore in current python versions.

**Comment by david-cortes-intel** [comment]
After the latest merges, is there something blocking this PR from applying this function to the current classes that can work with array API?

**Comment by icfaust** [comment]
> After the latest merges, is there something blocking this PR from applying this function to the current classes that can work with array API?

You are right on point with that comment.  Unfortunately none of the current estimators are fully array api compliant (realized I didn't include it while writing the first part of the docs), but a deluge of estimators should be ready for review using it in <1 month.

**Comment by david-cortes-intel** [comment]
> > After the latest merges, is there something blocking this PR from applying this function to the current classes that can work with array API?
> 
> You are right on point with that comment. Unfortunately none of the current estimators are fully array api compliant (realized I didn't include it while writing the first part of the docs), but a deluge of estimators should be ready for review using it in <1 month.

@icfaust Then I think it'd be better to hold off this PR until there are estimators ready to apply and test in the same PR.

**Comment by Vika-F** [comment]
Maybe a stupid question, but wouldn't it require checking the version of sklearn (1.3, 1.6) further in the code of each estimator to use the proper API?
Is it possible to decrease the number of sklearn versions check in the code of the sklearnex estimators?

**Review by ahuber21 — APPROVED** [review]

**Comment by icfaust** [comment]
> Maybe a stupid question, but wouldn't it require checking the version of sklearn (1.3, 1.6) further in the code of each estimator to use the proper API? Is it possible to decrease the number of sklearn versions check in the code of the sklearnex estimators?

That is a good question, this will take care of all issues related to array API support at class definition time (import time), meaning those additional checks throughout the codebase are all taken care of here: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/sklearnex/_device_offload.py#L118 in combination with https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/sklearnex/_utils.py#L31-L43 .  These will hide all of the sklearn version oddities related to it from the developer and user and are all done at import time, and makes all of the interfaces look like latest release sklearn with respect to tag interfaces.
